modules/kubernetes-addons/aws-vpc-cni/main.tf

```diff
@@ -1,32 +1,71 @@
+locals {
+  create_irsa = try(var.addon_config.service_account_role_arn == "", true)
+
+  cni_ipv6_policy = var.enable_ipv6 ? [aws_iam_policy.cni_ipv6_policy[0].arn] : []
+}
+
 resource "aws_eks_addon" "vpc_cni" {
   cluster_name             = var.addon_context.eks_cluster_id
-  addon_name               = local.addon_config["addon_name"]
-  addon_version            = local.addon_config["addon_version"]
-  resolve_conflicts        = local.addon_config["resolve_conflicts"]
-  service_account_role_arn = local.addon_config["service_account_role_arn"] == "" ? module.irsa_addon.irsa_iam_role_arn : local.addon_config["service_account_role_arn"]
-  preserve                 = local.addon_config["preserve"]
+  addon_name               = "vpc-cni"
+  addon_version            = try(var.addon_config.addon_version, null)
+  resolve_conflicts        = try(var.addon_config.resolve_conflicts, null)
+  service_account_role_arn = local.create_irsa ? module.irsa_addon[0].irsa_iam_role_arn : try(var.addon_config.service_account_role_arn, null)
+  preserve                 = try(var.addon_config.preserve, true)
+
   tags = merge(
-    var.addon_context.tags, local.addon_config["tags"],
-    { "eks_addon" = "vpc-cni" }
+    var.addon_context.tags,
+    try(var.addon_config.tags, {})
   )
-
-  depends_on = [module.irsa_addon]
 }
 
 module "irsa_addon" {
-  source                            = "../../../modules/irsa"
+  source = "../../../modules/irsa"
+
+  count = local.create_irsa ? 1 : 0
+
   create_kubernetes_namespace       = false
   create_kubernetes_service_account = false
-  kubernetes_namespace              = local.addon_config["namespace"]
-  kubernetes_service_account        = local.addon_config["service_account"]
-  irsa_iam_policies                 = local.irsa_iam_policies
+  kubernetes_namespace              = "kube-system"
+  kubernetes_service_account        = "aws-node"
   addon_context                     = var.addon_context
+  irsa_iam_policies = concat(
+    ["arn:${var.addon_context.aws_partition_id}:iam::aws:policy/AmazonEKS_CNI_Policy"],
+    local.cni_ipv6_policy,
+    try(var.addon_config.additional_iam_policies, [])
+  )
 }
 
 resource "aws_iam_policy" "cni_ipv6_policy" {
-  count       = var.enable_ipv6 ? 1 : 0
+  count = local.create_irsa && var.enable_ipv6 ? 1 : 0
+
   name        = "${var.addon_context.eks_cluster_id}-AmazonEKS_CNI_IPv6_Policy"
   description = "IAM policy for EKS CNI to assign IPV6 addresses"
+  path        = try(var.addon_context.irsa_iam_role_path, null)
   policy      = data.aws_iam_policy_document.ipv6_policy[0].json
-  tags        = var.addon_context.tags
+
+  tags = merge(
+    var.addon_context.tags,
+    try(var.addon_config.tags, {})
+  )
+}
+
+data "aws_iam_policy_document" "ipv6_policy" {
```

Comment:
move to data.tf for consistency?

Reply:
Lets not scatter code about and instead put it in close proximity of where its used. If I am looking at the codebase and I see a policy getting created, why not just look up/down a few lines instead of navigating to other files and searching? Same with locals - should be used sparingly and located close to where they are used. If we follow this pattern, would everything else then move into `resources.tf`?!

Reply:
Also, to reference a source other than my opinion - https://www.terraform.io/language/modules/develop/structure its not part of the recommended module structure. Typically its just `main.tf`, `variables.tf`, `outputs.tf` unless things get unwieldy, then you start looking at breaking out files

Reply:
10-4